mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp

```diff
@@ -0,0 +1,285 @@
+#define COLORIZE_DEADZONE 2.0f
+
+using namespace vgui;
+
+SpeedometerLabel::SpeedometerLabel(Panel *parent, const char *panelName, SpeedometerColorize_t colorizeType)
+    : Label(parent, panelName, ""), m_eColorizeType(colorizeType), m_pFlAlpha(nullptr),
+      m_pSzAnimationName(nullptr), m_bSupportsEnergyUnits(false), m_bDoneFading(false),
+      m_eUnitType(SPEEDOMETER_UNITS_UPS), m_pRangeList(nullptr), m_bDrawComparison(true), 
+      m_bSupportsSeparateComparison(true), m_cvarGravity("sv_gravity")
+{
+    CreateComparisonLabel();
+    Reset();
+}
+
+void SpeedometerLabel::SetVisible(bool bVisible)
+{
+    BaseClass::SetVisible(bVisible);
+    // parent's layout depends on the visiblity of this, so invalidate it
+    GetParent()->InvalidateLayout();
+}
+
+// creates a separate label for the comparison & pins it to this
+void SpeedometerLabel::CreateComparisonLabel()
+{
+    char name[BUFSIZELOCL];
+    Q_snprintf(name, sizeof(name), "%sComparison", GetName());
+    m_pComparisonLabel = new Label(GetParent(), name, "");
+    m_pComparisonLabel->SetName(name);
+    m_pComparisonLabel->SetPos(0, 0);
+    m_pComparisonLabel->PinToSibling(GetName(), PIN_TOPLEFT, PIN_TOPRIGHT);
+    m_pComparisonLabel->SetFont(GetFont());
+    m_pComparisonLabel->SetAutoTall(true);
+    m_pComparisonLabel->SetAutoWide(true);
+    m_pComparisonLabel->SetContentAlignment(a_center);
+    m_pComparisonLabel->SetPinCorner(PIN_TOPLEFT, 0, 0);
+}
+
+void SpeedometerLabel::SetCustomDiff(float diff)
+{
+    m_flDiff = diff;
+    m_bCustomDiff = true;
+}
+
+void SpeedometerLabel::ApplySchemeSettings(IScheme *pScheme)
+{
+    BaseClass::ApplySchemeSettings(pScheme);
+    m_NormalColor = GetSchemeColor("MOM.Speedometer.Normal", pScheme);
+    m_IncreaseColor = GetSchemeColor("MOM.Speedometer.Increase", pScheme);
+    m_DecreaseColor = GetSchemeColor("MOM.Speedometer.Decrease", pScheme);
+    m_pComparisonLabel->SetFont(GetFont());
+}
+
+void SpeedometerLabel::PerformLayout()
+{
+    BaseClass::PerformLayout();
+
+    // keep centered in parent
+    char szMain[BUFSIZELOCL];
+    GetText(szMain, BUFSIZELOCL);
+
+    HFont labelFont = GetFont();
+    int combinedLength = UTIL_ComputeStringWidth(labelFont, szMain);
+
+    if (m_eColorizeType == SPEEDOMETER_COLORIZE_COMPARISON_SEPARATE && m_bDrawComparison)
+    {
+        char szComparison[BUFSIZELOCL];
+        m_pComparisonLabel->GetText(szComparison, BUFSIZELOCL);
+        combinedLength += UTIL_ComputeStringWidth(labelFont, szComparison);
+    }
+
+    int xOffset = (GetParent()->GetWide() - combinedLength) / 2;
+    SetPos(xOffset, GetYPos());
+}
+
+void SpeedometerLabel::OnThink()
+{
+    if (HasFadeOutAnimation() && !m_bDoneFading)
+    {
+        SetAlpha(*m_pFlAlpha);
+        m_pComparisonLabel->SetAlpha(*m_pFlAlpha);
+
+        if (CloseEnough(*m_pFlAlpha, 0.0f, FLT_EPSILON))
+        {
+            m_bDoneFading = true;
+            Reset();
+        }
+    }
+}
```

Comment:
Always call base class for overrides. Early return for ease of reading.
```suggestion
void SpeedometerLabel::OnThink()
{
    BaseClass::OnThink();
    
    if (!HasFadeOutAnimation())
        return;
    
    if (m_bDoneFading)
        return;
    
    SetAlpha(*m_pFlAlpha);
    m_pComparisonLabel->SetAlpha(*m_pFlAlpha);
    if (CloseEnough(*m_pFlAlpha, 0.0f, FLT_EPSILON))
    {
        m_bDoneFading = true;
        Reset();
    }
}
```